scenarios/smart_trailer/applications/smart_trailer_application/src/main.rs

```diff
@@ -0,0 +1,252 @@
+
+    let response = client.get_subscription_info(request).await?;
+
+    Ok(response.into_inner())
+}
+
+/// Receive Trailer Weight updates.
+///
+/// # Arguments
+/// * `broker_uri` - The broker URI.
+/// * `topic` - The topic.
+async fn receive_trailer_weight_updates(
+    broker_uri: &str,
+    topic: &str,
+) -> Result<JoinHandle<()>, String> {
+    // Create a unique id for the client.
+    let client_id = format!("{MQTT_CLIENT_ID}-{}", Uuid::new_v4());
+
+    let create_opts = mqtt::CreateOptionsBuilder::new()
+        .server_uri(broker_uri)
+        .client_id(client_id)
+        .finalize();
+
+    let client = mqtt::Client::new(create_opts)
+        .map_err(|err| format!("Failed to create the client due to '{err:?}'"))?;
+
+    let receiver = client.start_consuming();
+
+    // Setup task to handle clean shutdown.
+    let ctrlc_cli = client.clone();
+    tokio::spawn(async move {
+        _ = signal::ctrl_c().await;
+
+        // Tells the client to shutdown consuming thread.
+        ctrlc_cli.stop_consuming();
+    });
+
+    // Last Will and Testament
+    let lwt = mqtt::MessageBuilder::new()
+        .topic("test")
+        .payload("Receiver lost connection")
+        .finalize();
+
+    let conn_opts = mqtt::ConnectOptionsBuilder::new_v5()
+        .keep_alive_interval(Duration::from_secs(30))
+        .clean_session(false)
+        .will_message(lwt)
+        .finalize();
+
+    client
+        .connect(conn_opts)
+        .map_err(|err| format!("Failed to connect due to '{err:?}"))?;
+
+    client
+        .subscribe(topic, mqtt::types::QOS_1)
+        .map_err(|err| format!("Failed to subscribe to topic {topic} due to '{err:?}'"))?;
+
+    // Copy topic for separate thread.
+    let topic_string = topic.to_string();
+
+    let sub_handle = tokio::spawn(async move {
+        for msg in receiver.iter() {
+            if let Some(msg) = msg {
+                // Here we log the message received. This could be expanded to parsing the message,
+                // Obtaining the weight and making decisions based on the weight
+                // For example, adjusting body functions or powertrain of the towing vehicle.
+                info!("{}", msg);
+            } else if !client.is_connected() {
+                if client.reconnect().is_ok() {
+                    client
+                        .subscribe(topic_string.as_str(), mqtt::types::QOS_1)
+                        .map_err(|err| format!("Failed to subscribe to topic {topic_string} due to '{err:?}'"))?;
+                } else {
+                    break;
+                }
+            }
+        }
+
+        if client.is_connected() {
+            debug!("Disconnecting");
+            client
+                .unsubscribe(topic_string.as_str())
+                .map_err(|err| format!("Error unsubscribing: {err}"))?;
+            client
+                .disconnect(None)
+                .map_err(|err| format!("Error disconnecting: {err}"))?;
+        }
+    });
```

Comment:
To make use of Will's suggestion, add an `Ok(())` after the if statement. This is needed because futures determine their return type based on whats going on in the function.